BJS2-78085 ai spelling

Fagan89 opened the pull request:
не доделаны тесты

Commit:
BJS2-78085
Добавил запрос в базу, получение неопубликованных ростов

Commit:
BJS2-78085

Commit:
BJS2-78085
1. сделан сервис для обращения к LanguageTool
2. сервис обработки ответа от LanguageTool

Commit:
Merge branch 'refs/heads/hippogriff-master-stream10' into feature/BJS2-78085-ai-spelling

# Conflicts:
#	src/main/resources/application.yaml

Commit:
BJS2-78085
1. исправил LanguageToolService

Aleck210popov commented on src/main/java/faang/school/postservice/repository/PostRepository.java:
```diff
@@ -48,4 +48,12 @@ public interface PostRepository extends JpaRepository<Post, Long> {
+    @Query(nativeQuery = true, value = """
+            SELECT * FROM POST post
+            WHERE post.published = false AND post.deleted = false
+            """)
```
Можно published и deleted передавать в метод как параметры. Тогда нативку не придется писать.

Но это скорее как альтернатива

Aleck210popov commented on src/main/java/faang/school/postservice/config/context/AsyncConfig.java:
```diff
@@ -0,0 +1,26 @@
+public class AsyncConfig {
+
+    @Bean
+    public Executor customExecutor() {
```
Возможно это не последний Executor на проекте, нужно бы более осмысленное название

Aleck210popov commented on src/main/java/faang/school/postservice/service/post/PostCorrecter.java:
```diff
@@ -0,0 +1,82 @@
+        log.info("Finished spellcheck job");
+    }
+
+    @Async("customExecutor")
```
Не сработает при вызове из того же бина

Aleck210popov commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -89,6 +89,11 @@ public List<Post> getAllPublishedByProjectId(Long projectId) {
         return postRepository.findPublishedByProjectId(projectId);
     }
 
+    @Transactional
```
readOnly = true

Aleck210popov commented on src/main/java/faang/school/postservice/service/post/PostCorrecter.java:
```diff
@@ -0,0 +1,82 @@
+                    String correctedContent = applyCorrected(content, response);
+                    post.setContent(correctedContent);
+                    postRepository.save(post);
+                    log.info("post text ID: {} updated", post.getId());
```
Может есть смысл эти 3 строки перенести в сервис, а то получается что джоба работает сразу с двумя слоями. Сервис и репозиторий.

Commit:
BJS2-78085
1. исправил LanguageToolService

Commit:
BJS2-78085
1. Добавлены тесты

Commit:
BJS2-78085
1. Добавлены тесты

Commit:
Merge branch 'hippogriff-master-stream10' into feature/BJS2-78085-ai-spelling

# Conflicts:
#	src/main/resources/application.yaml

Commit:
BJS2-78085
пофиксил код, добавил retry

Commit:
BJS2-78085
поправил yaml

Commit:
BJS2-78085
fix

20cognosce commented on src/main/java/faang/school/postservice/config/context/AppConfig.java:
```diff
@@ -0,0 +1,14 @@
+public class AppConfig {
+
+    @Bean
+    public RestTemplate restTemplate() {
```
имя бину давай, у тебя не один рест темплейт на все внешние сервисы

20cognosce commented on src/main/java/faang/school/postservice/config/executor/ConfiguredExecutorService.java:
```diff
@@ -18,6 +19,7 @@ public class ConfiguredExecutorService {
 
 
     @Bean(name = "taskExecutor")
+    @Primary
```
с чего возникла такая необходимость?

20cognosce commented on src/main/resources/application.yaml:
```diff
@@ -65,4 +65,10 @@ moderation:
+  correction:
+    url: "https://api.languagetool.org/v2/check"
+    language: "auto"
+    cron: "0 */30 * * * *"
```
каждые 30 минут? пиши комментарий к cron-ам

20cognosce commented on src/main/java/faang/school/postservice/service/PostService.java:
```diff
@@ -89,9 +91,21 @@ public List<Post> getAllPublishedByProjectId(Long projectId) {
+        post.setContent(correctedContent);
+        postRepository.save(post);
+        log.info("post text ID: {} updated", post.getId());
+    }
```
как-то странно видеть это в PostService, где методы для работы с постами для юзеров
Эти методы ни один юзер никогда использовать не будет

логичнее в другой сервис это вынести

Fagan89 replied:
да так будет логичней. 
у меня была другая логика, что все метода с транзакцией лучше тут реализовывать.Буду иметь ввиду

20cognosce commented on src/main/java/faang/school/postservice/repository/PostRepository.java:
```diff
@@ -48,4 +48,10 @@ public interface PostRepository extends JpaRepository<Post, Long> {
+            SELECT * FROM POST post
+            WHERE post.published = false AND post.deleted = false
+            """)
+    List<Post> findAllUnpublishedPosts();
```
2 инстанса одновременно запустят джобы по крону и тут один из твоих запросов положит джобу

20cognosce commented on src/main/java/faang/school/postservice/jobs/CorrectorUnpublishPosts.java:
```diff
@@ -0,0 +1,38 @@
+
+        List<Post> posts = postService.getAllUnpublishedPost();
+        List<CompletableFuture<Void>> futures = posts.stream()
+                .map(postCorrecter::correctingContentPost)
```
а где транзакция(-и)?
и если у нас 100_000 постов, ты на каждый из них отдельный поток будешь выделять?

ты должен из базы батчами по 100-1000 постов С ЛОКОМ доставать в одной транзакции их все корректировать в этой же транзакции сохранять

батчи уже можно распараллеливать, но не отдельные посты

20cognosce commented on src/main/java/faang/school/postservice/service/post/PostCorrecter.java:
```diff
@@ -0,0 +1,57 @@
+    private final LanguageToolService languageToolService;
+
+    @Async("taskExecutor")
+    public CompletableFuture<Void> correctingContentPost(Post post) {
```
зачем тебе тут вообще CompletableFuture нужен, что кому ты возвращаешь? Пустоту?

20cognosce commented on src/main/java/faang/school/postservice/service/languagetool/LanguageToolService.java:
```diff
@@ -0,0 +1,52 @@
+        formData.add("language", language);
+
+        log.debug("Sending request to LanguageTool for text: {}", originalText);
+        return CompletableFuture
+                .supplyAsync(
```
зачем и тут еще поток..

зашли в джобу - новый поток
каждый пост в джобе - новый поток
каждый запрос по каждому посту в джобе - новый поток

Для чего это всё?

Commit:
Merge branch 'refs/heads/hippogriff-master-stream10' into feature/BJS2-78085-ai-spelling

# Conflicts:
#	src/main/java/faang/school/postservice/service/post/PostService.java
#	src/main/resources/application.yaml
#	src/test/java/faang/school/postservice/service/post/PostServiceTest.java

Commit:
BJS2-78085
1. всe переделал
2. исправил замечания
3. добавил к таблице пост, еще одну колонку с исправленным текстом
4. Джоба запускается асинхронно и далее асинхронка запускает работу по каждому батчу

Commit:
BJS2-78085
конфликты

Commit:
BJS2-78085
1. добавил дополнительный, что бы джоба запускалась на кастомном Executor

20cognosce commented on src/main/java/faang/school/postservice/service/post/PostService.java:
```diff
@@ -114,6 +116,17 @@ public List<Post> getAllPublishedByProjectId(Long projectId) {
         return postRepository.findPublishedByProjectId(projectId);
     }
 
+    @Transactional
+    @Async("executorForPostService")
```
вместе не использовать

Fagan89 replied:
принял, но это работает, проверил ))

20cognosce reviewed (approved):
норм
